src/composables/use-fetch-state.ts

```diff
@@ -0,0 +1,111 @@
+import { computed, reactive, Ref, ref, watch } from '@nuxtjs/composition-api'
+
+import type { FetchState } from '~/store/types'
+
+/* Constants */
+
+/**
+ * Statuses of requests:
+ * - `idle`: the current request has never been sent yet, for example,
+ * on the first app load, or after a change in search filters or search term.
+ * - `fetching`: the request was sent, but no response was received yet.
+ * - `success`: a successful response was received for the current request.
+ * - `finished`: for multi-page requests, this is true when no more pages are left.
+ * - `error`: an error response was received.
+ */
+const statuses = {
+  IDLE: 'idle',
+  FETCHING: 'fetching',
+  SUCCESS: 'success',
+  ERROR: 'error',
+  FINISHED: 'finished',
+} as const
+
+type Status = typeof statuses[keyof typeof statuses]
+/**
+ * When one of these statuses is received, the fetchError is set to `null` in a watcher.
+ */
+const nonErrorStatuses: Status[] = [
+  statuses.IDLE,
+  statuses.FETCHING,
+  statuses.SUCCESS,
+]
+/**
+ * With these statuses, it is possible to send the same request for a new page.
+ * This can help debounce requests and prevent racing states.
+ */
+const canFetchAgainStatuses: Status[] = [statuses.IDLE, statuses.SUCCESS]
+
+/* Composable */
+
+export const useFetchState = (initialState = statuses.IDLE) => {
+  const fetchStatus: Ref<Status> = ref(initialState)
+  const fetchError: Ref<string | null> = ref(null)
+
+  watch(fetchStatus, () => {
+    if (nonErrorStatuses.includes(fetchStatus.value)) {
+      fetchError.value = null
+    }
+  })
+  const reset = () => {
+    fetchStatus.value = statuses.IDLE
+  }
+  const startFetching = () => {
+    fetchStatus.value = statuses.FETCHING
+  }
+  /**
+   * Called when any response is received. For an error response,
+   * pass the errorMessage string as a parameter.
+   * @param errorMessage - the message to show for response error.
+   */
+  const endFetching = ({
+    errorMessage = undefined,
+  }: { errorMessage?: string } = {}) => {
```

Comment:
Just wondering why this is an object parameter rather than a plain string param. It seems like it'd be simpler as a string param:

```suggestion
  const endFetching = (errorMessage?: string) => {
```

Do we anticipate there being other things needing to be passed here?